encoder.py

```diff
@@ -0,0 +1,106 @@
+from torch.hub import load_state_dict_from_url
+
+
+class ImageTorchEncoder(Executor):
+    """
+    :class:`ImageTorchEncoder` encodes ``Document`` content from a ndarray,
+    potentially B x (Channel x Height x Width) into a ndarray of `B x D`.
+    Where B` is the batch size and `D` is the Dimension.
+    Internally, :class:`ImageTorchEncoder` wraps the models from `
+    `torchvision.models`.
+    https://pytorch.org/docs/stable/torchvision/models.html
+    :param model_name: the name of the model. Supported models include
+        ``resnet18``, ``alexnet``, `squeezenet1_0``,  ``vgg16``,
+        ``densenet161``, ``inception_v3``, ``googlenet``,
+        ``shufflenet_v2_x1_0``, ``mobilenet_v2``, ``resnext50_32x4d``,
+        ``wide_resnet50_2``, ``mnasnet1_0``
+    :param pool_strategy: the pooling strategy. Options are:
+        - `None`: Means that the output of the model will be the 4D tensor
+            output of the last convolutional block.
+        - `mean`: Means that global average pooling will be applied to the
+            output of the last convolutional block, and thus the output of
+            the model will be a 2D tensor.
+        - `max`: Means that global max pooling will be applied.
+    :param channel_axis: The axis of the color channel, default is 1
+    :param args:  Additional positional arguments
+    :param kwargs: Additional keyword arguments
+    """
+    DEFAULT_TRAVERSAL_PATH = ['r']
+
+    def __init__(
+        self,
+        model_name: str = 'mobilenet_v2',
+        pool_strategy: str = 'mean',
+        channel_axis: int = 1,
+        load_pre_trained_from_path: Optional[str] = None,
+        default_traversal_path: Optional[List[str]] = None,
+        *args,
+        **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self.default_traversal_path = self.DEFAULT_TRAVERSAL_PATH if default_traversal_path is None\
+            else default_traversal_path
+
+        self.channel_axis = channel_axis
+        # axis 0 is the batch
+        self._default_channel_axis = 1
+        self.model_name = model_name
+        if pool_strategy not in ('mean', 'max'):
+            raise NotImplementedError(f'unknown pool_strategy: {self.pool_strategy}')
+        self.pool_strategy = pool_strategy
+
+        if load_pre_trained_from_path:
+            # set env var as described in https://pytorch.org/vision/stable/models.html
+            model = getattr(models, self.model_name)(pretrained=False)
+            model.load_state_dict(torch.load(load_pre_trained_from_path))
+
+
+        model = getattr(models, self.model_name)(pretrained=True)
+        self.model = model.features.eval()
+        self.model.to(torch.device('cpu'))
+        if self.pool_strategy is not None:
+            self.pool_fn = getattr(np, self.pool_strategy)
+
+    def _get_features(self, content):
+        return self.model(content)
+
+    def _get_pooling(self, feature_map: 'np.ndarray') -> 'np.ndarray':
```

Comment:
One big note about this `executor`. This poling `strategy` is not the strategy that should be followed by avery model. For instance `Resnet` based models ave a specific `GlobalAveragePooling` layer to extract

Reply:
We found a temporary solution for now and left a TODO mark to improve this in the future.

https://github.com/jina-ai/executor-image-torch-encoder/pull/1/commits/2838a478c695643f3d6f6ff45ef4ad33de0c02b4

Reply:
But for those who have avgPooling we can already apply it right?

Reply:
What about using the `AdaptiveAveragePooling2D` internally and trying to always extract the `features` from the models?

Reply:
I do not understand what u mean?

Reply:
Sorry, I meant that for all models we try to extract the feature layer (prior to pooling) and then add our own pooling strategy (e.g. `AdaptiveAveragePooling2D`) instead of differentiating which models have the pooling layer and which don't. Seem more consistent to me.

Reply:
Yes, that is the ideal

Reply:
But u told me they do not all have features submodel right?

Reply:
they do, it is just named differently (it it either `features` or `fc` and for one model it is `layers`)

Reply:
This code attempts to always find the features submodule so that we can consistently apply the pooling ourselves.
```
if hasattr(model, 'features'):
    return model.features.eval()
elif hasattr(model, 'fc'):
    return model.eval()
elif hasattr(model, 'layers'):
   return model.layers.eval()
else:
    raise ValueError(f'Model {model.__class__.__name__} is currently not supported by the ImageTorchEncoder')
```